consensus/polybft/validators_snapshot.go

```diff
@@ -38,58 +36,55 @@ func (v *validatorsSnapshotCache) GetSnapshot(blockNumber uint64, parents []*typ
 	v.lock.Lock()
 	defer v.lock.Unlock()
 
-	epochNumber := getEpochNumber(blockNumber, v.epochSize)
-	if !isEndOfPeriod(blockNumber, v.epochSize) && epochNumber > 0 {
-		// snapshot is retrieved for the previous epoch
-		// since the last block of the previous epoch contains validator set for the current epoch
-		epochNumber--
-	}
-
-	v.logger.Trace("Retrieving snapshot started...", "Block", blockNumber, "Epoch", epochNumber)
+	v.logger.Trace("Retrieving snapshot started...", "Block", blockNumber)
 
 	var snapshot AccountSet
 
-	currentEpochNumber := epochNumber
+	lastBlockWithSnapshot := blockNumber
 
-	for ; ; currentEpochNumber-- {
-		snapshot = v.snapshots[currentEpochNumber]
+	for ; ; lastBlockWithSnapshot-- {
+		snapshot = v.snapshots[lastBlockWithSnapshot]
 		if snapshot != nil {
-			v.logger.Trace("Found snapshot in memory cache", "Epoch", currentEpochNumber)
+			v.logger.Trace("Found snapshot in memory cache", "Block", lastBlockWithSnapshot)
 
 			break
 		}
 
-		dbSnapshot, err := v.state.getValidatorSnapshot(currentEpochNumber)
+		dbSnapshot, err := v.state.getValidatorSnapshot(lastBlockWithSnapshot)
 		if err != nil {
 			return nil, fmt.Errorf(
-				"failed to retrieve validators snapshot from the database for epoch=%d: %w",
-				currentEpochNumber,
+				"failed to retrieve validators snapshot from the database for block=%d: %w",
+				lastBlockWithSnapshot,
 				err,
 			)
 		}
 
 		if dbSnapshot != nil {
 			snapshot = dbSnapshot
-			v.snapshots[currentEpochNumber] = dbSnapshot.Copy()
-			v.logger.Trace("Found snapshot in database", "Epoch", currentEpochNumber)
+			v.snapshots[lastBlockWithSnapshot] = dbSnapshot.Copy()
+			v.logger.Trace("Found snapshot in database", "Block", lastBlockWithSnapshot)
 
 			break
 		}
 
-		if currentEpochNumber == 0 {
+		if lastBlockWithSnapshot == 0 {
 			// prevent underflow of uint64
 			break
 		}
 	}
 
-	if snapshot != nil && currentEpochNumber == epochNumber {
+	if snapshot != nil && lastBlockWithSnapshot == blockNumber {
+		if err := v.cleanup(); err != nil {
```

Comment:
I think we can get rid of cleanup. 
marshalled validator size = 20b+ 32b*4+8b = 156b
validator set size = 15kb
Snapshot storage/year(for epoch=1d) = 15kb*365 = 5.5mb

Reply:
> Snapshot storage/year(for epoch=1d) = 15kb * 365 = 5.5mb

If validator set changes more often than once/day, this can increase.

IMO we should keep cleanup logic, since it doesn't introduce too much complexity.

Reply:
I agree. We can increase the number of snapshots we keep in storage though.

Reply:
if our epoch length will be 30m -  our data consumtion will be:
15kb * 365 * 48 = 264mb/year.
I'm not against cleaning up in general, but I'm confused that we want to remove our knowlidge about epoch's validator set. It's not taking much space.

Reply:
> 15kb * 365 * 48 = 264mb/year.

Can you just elaborate what 48 represents in this formula? It has to do something with snapshot storing frequency in a single day (so 48 times I'd say).

This PR introduces persistence of snapshot per blocks (instead of per epoch) therefore it almost certainly is going to be larger, since we are retrieving validators snapshot per each block when verifying signatures. 
If we assume that block time is 2 s (which was by default on v3), it means that during a single day we should produce 43200. So my formula looks like this (please check if this makes sense):
15kb * 365 * 43200 = ~236k mb/year, which is quite a lot :slightly_smiling_face:.
Whenever we construct a snapshot we store it in the cache and db. Block within the epoch will have the redundant data (i.e. same validator sets).

Even if we preserve cleanup logic, we can always reconstruct snapshot based on last persisted snapshot and deltas.

Reply:
I started woeking on that task for storing snapshots by validator set hash, so it won't be by block anymore.

Reply:
I would say let's decide about cleanup on that following PR. For now, let's keep it like this.